### PR TITLE
Fix: Timeout/OOM for some subgraphs in snapshot hca_prod_20201120_dcp2___20210302 (#2868)

### PR DESCRIPTION
**Pull request by noah-aviel-dove**
Author

- [x] PR title references issue
- [x] Title of main commit references issue
- [x] PR is linked to Zenhub issue

Author (reindex)

- [x] Added `r` tag to commit title                         <sub>or this PR does not require reindexing</sub>
- [x] Added `reindex` label to PR                           <sub>or this PR does not require reindexing</sub>

Author (freebies & chains)

- [x] Freebies are blocked on this PR                       <sub>or there are no freebies in this PR</sub>
- [x] Freebies are referenced in commit titles              <sub>or there are no freebies in this PR</sub>
- [x] This PR is blocked by previous PR in the chain        <sub>or this PR is not chained to another PR</sub>
- [x] Added `chain` label to the blocking PR                <sub>or this PR is not chained to another PR</sub>

Author (upgrading)

- [x] Documented upgrading of deployments in UPGRADING.rst  <sub>or this PR does not require upgrading</sub>
- [x] Added `u` tag to commit title                         <sub>or this PR does not require upgrading</sub>
- [x] Added `upgrade` label to PR                           <sub>or this PR does not require upgrading</sub>

Author (requirements, before every review)

- [x] Ran `make requirements_update`                        <sub>or this PR leaves requirements*.txt, common.mk and Makefile untouched</sub>
- [x] Added `R` tag to commit title                         <sub>or this PR leaves requirements*.txt untouched</sub>
- [x] Added `reqs` label to PR                              <sub>or this PR leaves requirements*.txt untouched</sub>

Author (before every review)

- [x] `make integration_test` passes in personal deployment <sub>or this PR does not touch functionality that could break the IT</sub>
- [x] Rebased branch on `develop`, squashed old fixups

Primary reviewer (after approval)

- [ ] Commented on demo expectations                        <sub>or labelled PR as `no demo`</sub>
- [x] Decided if PR can be labeled `no sandbox`
- [x] PR title is appropriate as title of merge commit
- [x] Moved ticket to Approved column
- [x] Assigned PR to an operator

Operator (before pushing merge the commit)

- [x] Checked `reindex` label and `r` commit title tag
- [ ] Checked that Demo expectations are clear              <sub>or PR is labeled as `no demo`</sub>
- [x] Rebased and squashed branch
- [x] Sanity-checked history
- [x] Pushed PR branch to Github
- [x] Branch pushed to Gitlab and build passes in `sandbox` <sub>or added `no sandbox` label</sub>
- [x] Started reindex in `sandbox`                          <sub>or this PR does not require reindexing `sandbox`</sub>
- [x] Checked for failures in `sandbox`                     <sub>or this PR does not require reindexing `sandbox`</sub>
- [x] Added PR reference to merge commit title
- [x] Collected commit title tags in merge commit title
- [x] Moved linked issue to Merged column
- [ ] Pushed merge commit to Github

Operator (after pushing the merge commit)

- [ ] Moved freebies to Merged column                       <sub>or there are no freebies in this PR</sub> 
- [ ] Shortened the PR chain                                <sub>or this PR is not the base of another PR</sub>
- [ ] Verified that `N reviews` labelling is accurate
- [ ] Pushed merge commit to Gitlab                         <sub>or this changes can be pushed later, together with another PR</sub>
- [ ] Deleted PR branch from Github and Gitlab

Operator (reindex) 

- [ ] Purchased BQ slot committment in `dev`                <sub>or this PR does not require reindexing `dev`</sub>
- [ ] Started reindex in `dev`                              <sub>or this PR does not require reindexing `dev`</sub>
- [ ] Deleted BQ slot committment in `dev`                  <sub>or this PR does not require reindexing `dev`</sub>
- [ ] Checked for failures in `dev`                         <sub>or this PR does not require reindexing `dev`</sub>
- [ ] Purchased BQ slot committment in `prod`               <sub>or this PR does not require reindexing `prod`</sub>
- [ ] Started reindex in `prod`                             <sub>or this PR does not require reindexing `prod`</sub>
- [ ] Deleted BQ slot committment in `prod`                 <sub>or this PR does not require reindexing `prod`</sub>
- [ ] Checked for failures in `prod`                        <sub>or this PR does not require reindexing `prod`</sub>

Operator

- [ ] Unassigned PR

### CODE REVIEW

**Comment by noah-aviel-dove** [comment]
Can't run `make requirements_update` while the `hca_metadata_api` req is pinned to a specific commit while https://github.com/DataBiosphere/hca-metadata-api/issues/36 is pending. Presumably a new release will be made for that repo before final approval of this PR.

File: src/azul/plugins/repository/tdr/__init__.py [hunk 1]
```diff
@@ -546,14 +560,20 @@ def extract_field(field: attr.Attribute) -> Tuple[str, Any]:
 
 class TDRBundle(Bundle[TDRSourceRef]):
 
-    def add_entity(self, entity_key: str, entity_type: EntityType, entity_row: BigQueryRow) -> None:
+    def add_entity(self,
+                   entity_key: str,
```
Comment (hannes-ucsc) [inline src/azul/plugins/repository/tdr/__init__.py:564]:
```suggestion
                   *,
                   entity_key: str,
```

File: terraform/sqs.tf.json.template.py [hunk 2]
```diff
@@ -12,15 +12,19 @@
...
-                            "deadLetterTargetArn": "${aws_sqs_queue.%s.arn}"
-                                                   % config.unqual_notifications_queue_name(fail=True)
-                        })
+                    **{
+                        config.unqual_notifications_queue_name(retry=retry): {
+                            "name": config.notifications_queue_name(retry=retry),
+                            "visibility_timeout_seconds": config.contribution_lambda_timeout(retry=retry) + 10,
+                            "message_retention_seconds": 24 * 60 * 60,
+                            "redrive_policy": json.dumps({
+                                "maxReceiveCount": 10,  # TODO increase for retry?
```
Comment (hannes-ucsc) [inline terraform/sqs.tf.json.template.py:21]: Same as for aggregates: 1 for the first queue, 9 for the second, totaling 10 attempts as before.

File: src/azul/__init__.py [hunk 3]
```diff
@@ -754,6 +756,11 @@ def es_refresh_interval(self) -> int:
         """
         return 1
 
+    @property
+    def max_chunk_size(self) -> int:
```
Comment (hannes-ucsc) [inline src/azul/__init__.py:760]: Could just be a class attribute

**Review by hannes-ucsc — CHANGES_REQUESTED** [review]

**Review by hannes-ucsc — APPROVED** [review]

**Comment by hannes-ucsc** [comment]
```
$ python scripts/manage_queues.py list
2021-03-19 08:33:16,493 INFO    MainThread: Listing queues

Queue Name                          Messages Available  Messages In Flight  Messages Delayed 

2021-03-19 08:33:16,507 INFO    MainThread: Found credentials in environment variables.
azul-notifications-prod                     0                   0                  0         
azul-notifications_retry-prod               0                   0                  0         
azul-tallies-prod.fifo                      0                   0                  0         
azul-tallies_retry-prod.fifo                0                   0                  0         
azul-tallies_fail-prod.fifo                 0                   0                  0         
azul-notifications_fail-prod                2                   0                  0         
```

**Comment by hannes-ucsc** [comment]
```json
{
    "queue": "https://queue.amazonaws.com/542754589326/azul-notifications_fail-prod",
    "messages": [
        {
            "MessageId": "7efa3531-e49a-437c-a035-14f2e08c43d5",
            "ReceiptHandle": "AQEB1blRz8raFv8xJ9hSAb8nJKDx2ZDRsr/2cUDzPIt7ec7E2MC3xMU6SuoaSSUGYu5vJHwqxmHwk7A9aKmiKAhddSddWngHSllst93/wpjdHDJXP3LOWgccQMUO7r5Dg2yJ+j3ibd1uitR1c+GK2WvjEoFUO5Tg9T1aWt+4QbYPtq50gwv6qb6RIun6vm6av5ThptZ4WXfdDX00Jlsru42Qa3tQWnWa26H1tEq7394LtqxKys9hA8+0n8z7NcYhj9xXirz8d3ykOcwM+1a8P3rRD6rioN3SKvzg2Jtzmf+y0XKpXEZzRdtA7babE7rtIYwIKaPNZy8X1ljto9bN2qK6zHWqiq/IvCjPDN1WNvn1bAXt4W+hYK0VtdTfm5v7CIADIZ4EwJUFpEACmTcyUqSVfLm0fuFLX1hOroQUmZW6fdk=",
            "MD5OfBody": "0efdea6be8af93aacf8c690469fe27a7",
            "Body": {
                "action": "add",
                "notification": {
                    "source": {
                        "id": "4c9d9b2d-a469-40bc-b79b-c3465c154675",
                        "name": "tdr:broad-datarepo-terra-prod-hca2:snapshot/hca_prod_20201120_dcp2___20210315"
                    },
                    "query": {},
                    "subscription_id": "cafebabe-feed-4bad-dead-beaf8badf00d",
                    "transaction_id": "2a2a3597-722e-4032-86d0-e6ec444c1a44",
                    "match": {
                        "bundle_uuid": "8d455ae8-b82f-5e80-ad8c-b482ec764e23",
                        "bundle_version": "2021-02-22T16:08:23.715288Z"
                    }
                },
                "catalog": "dcp3"
            },
            "Attributes": {
                "SenderId": "AROAX4XVRZKHHXD45CIX7:azul-indexer-prod-contribute",
                "ApproximateFirstReceiveTimestamp": "1616133143062",
                "ApproximateReceiveCount": "11",
                "SentTimestamp": "1616133143062"
            },
            "_Attributes": {
                "SentTimestamp": "2021-03-18T22:52:23.062000-07:00",
                "ApproximateFirstReceiveTimestamp": "2021-03-18T22:52:23.062000-07:00"
            }
        },
        {
            "MessageId": "0ea40af5-8a43-4493-83c5-32ddb89518e1",
            "ReceiptHandle": "AQEBN8gplyQMTYU0lyvXgJRyn5VUbu4FKvHSBneFiP6ld1X8GCeK+2xUzs7KD7jCLeS7nh8iB3SMcwbeboisQP1sGKCD2mWWT4iCDMBt3iyzPRQY/mVXCC99mWz4ww8hZVPrv7HSvpNwdh0gnrJp9rah6U8E59AU6IVEENDR5y2teJ7ScbXpp3Fhk/bKXlurDJsGE3ky2Y3nxiQgsnlD2jzFhiJuQodz/KwvDcxKJ69Jk81zOR/3aZduk/OusyErchsPww6dJtuBr0TXPgfJVXI2AA2VCLKaGfGJqSJrYu7JlDzjbO1ZO33zPPKCz3wTul4soKRCzJBwGQiyUwC8W9f1uYp7c/TVEJLnUk15DfOdralhf8o8Qb+i4TS6s+OOPlF3SD3hut0o/fLcqXdOXrpJTgSg+pKROE/W3OVJKVK4qsQ=",
            "MD5OfBody": "18fe4df84343af816f7a7e53a74a827f",
            "Body": {
                "action": "add",
                "notification": {
                    "source": {
                        "id": "4c9d9b2d-a469-40bc-b79b-c3465c154675",
                        "name": "tdr:broad-datarepo-terra-prod-hca2:snapshot/hca_prod_20201120_dcp2___20210315"
                    },
                    "query": {},
                    "subscription_id": "cafebabe-feed-4bad-dead-beaf8badf00d",
                    "transaction_id": "653948e7-28b3-45c7-9013-c7687cf7323a",
                    "match": {
                        "bundle_uuid": "5b309e71-eb46-597a-bfdc-638d123e4eb5",
                        "bundle_version": "2021-02-22T16:08:23.715288Z"
                    }
                },
                "catalog": "dcp3"
            },
            "Attributes": {
                "SenderId": "AROAX4XVRZKHHXD45CIX7:azul-indexer-prod-contribute",
                "ApproximateFirstReceiveTimestamp": "1616133139772",
                "ApproximateReceiveCount": "11",
                "SentTimestamp": "1616133139772"
            },
            "_Attributes": {
                "SentTimestamp": "2021-03-18T22:52:19.772000-07:00",
                "ApproximateFirstReceiveTimestamp": "2021-03-18T22:52:19.772000-07:00"
            }
        }
    ]
}
```

**Comment by hannes-ucsc** [comment]
The above failures are tracked in #2867, the failures from #2868 are gone, which proves that the issue is resolved. No demo.

**Pull request merged**